### ISSUE TITLE
🐛 Invalid floatDiffInDays calculate

### ISSUE THREAD

**Issue by pnlinh-it** [briannesbitt/Carbon#2531]
Hello,

I encountered an issue with the following code:
```php
$now = new Carbon('2019-07-29 14:02:54.000000');

// Just change 29 -> 28 and 000000 -> 000001
$then1 = new Carbon('2019-07-28 14:02:54.000001');

// Just change 29 -> 27 and 000000 -> 000001
$then2 = new Carbon('2019-07-27 14:02:54.000001');

echo $now->floatDiffInDays($then1); // 0.99999999998843

echo PHP_EOL;

echo $now->floatDiffInDays($then2); // 2.9999999999884 
```

Carbon version: **2.55.2**

PHP version: **7.4**

I expected to get:

```php
echo $now->floatDiffInDays($then2); // 1.9999999999884 
```

But I actually get:

```php
echo $now->floatDiffInDays($then2); // 2.9999999999884 
```

Thanks!

**Comment by kylekatarnls** [comment]
For the record, this is actually a PHP bug that was fixed in PHP 8.1:
https://3v4l.org/29lef

I will check if it's worth taking the risk of trying a work-around here.

I recommend to upgrade to PHP 8.1.1 (current PHP stable version).

**Comment by pnlinh-it** [comment]
@kylekatarnls 
Thank you for your support 🙇 . I will upgrade to PHP 8.1.

**Comment by pnlinh-it** [comment]
@kylekatarnls I would like to share another case hope that will help you:

```php
Carbon\Carbon::parse('2022-01-04 13:32:30.628030')->floatDiffInDays(Carbon\Carbon::parse('2022-01-11 13:32:30.000000'))
// => 7.9999927311343
Carbon\Carbon::parse('2022-01-04 13:32:52.900330')->floatDiffInDays(Carbon\Carbon::parse('2022-01-11 13:32:30.000000'))
// => 6.9997349498843
```

**Comment by pnlinh-it** [comment]
@kylekatarnls

Hello, it's me again. After upgrading to PHP 8.1, I would like to confirm that it's working now. Thank you for your support!

**Comment by Brenneisen** [comment]
Hi! I am not sure if this is part of the same problem:

<img width="581" alt="Bildschirmfoto 2022-01-17 um 22 34 05" src="https://user-images.githubusercontent.com/5833477/149839803-dc9d01ac-3a50-4506-9b04-db73091ebb6f.png">
<img width="584" alt="Bildschirmfoto 2022-01-17 um 22 34 12" src="https://user-images.githubusercontent.com/5833477/149839806-e71370e9-4e10-4ed5-bbd2-f36ad86afa6f.png">

**Comment by kylekatarnls** [comment]
Hello.

It's related to the fix made in PHP 8.1.

I know at a first sight it looks like PHP 8.0 is correct and 8.1 has a bug, but indeed no.

See this example:
https://3v4l.org/KvCQs

```php
$frozenNow = new DateTimeImmutable();
$now = $frozenNow->modify('-2 day');
$then = $frozenNow->modify('-5 day');

var_dump($now->diff($then)->days);
```

The only difference is that now has been frozen and both "now" and "then" use the same same moment as a reference before the -X days modification.

In your code, it's not the case, you actually do:

now = (moment T) - 2 days
then = (moment T + few µs) - 5 days

So you don't have 3 days as a difference, you have 3 days - few µs (the time needed by PHP to process the instruction. No code execution can be instantaneous.

**Comment by Brenneisen** [comment]
@kylekatarnls That makes sense! Thank you!
